### PR TITLE
metrics: change metadata vector to chunked_fifo

### PR DESCRIPTION
**Pull request by NyaliaLui**
Upon resizing, a std::vector will allocate a large piece of contiguous memory which may become fragmented over time. This can lead to oversized allocations on the metadata vector occurring at the call to vector.emplace_back():

```cpp
void impl::update_metrics_if_needed() {
if (_dirty) {
   _metadata = ::seastar::make_shared<metric_metadata>();

   auto mt_ref = ::seastar::make_shared<metric_metadata>();
   auto &mt = *(mt_ref.get());
   mt.reserve(_value_map.size());
   _current_metrics.resize(_value_map.size());
   size_t i = 0;
   for (auto&& mf : _value_map) {
       metric_metadata_vector metrics;
       _current_metrics[i].clear();
       for (auto&& m : mf.second) {
           if (m.second && m.second->is_enabled()) {
               metrics.emplace_back(m.second->info()); // Backtrace shows overalloc occurs here
               _current_metrics[i].emplace_back(m.second->get_function());
           }
       }
       if (!metrics.empty()) {
           mt.emplace_back(metric_family_metadata{mf.second.info(), std::move(metrics)});
           i++;
       }
   }
   _current_metrics.resize(i);
   _metadata = mt_ref;
   _dirty = false;
}
}
```

Therefore, this PR switches the metadata vector with a chunked_fifo.

Fixes: https://github.com/scylladb/seastar/issues/1844

Changes from force-push `d8662a3`:
- Split the patches into two commits
- Add more explanation to why `chunked_fifo::basic_iterator` needs a default ctor

Changes from force-push `55d6663`:
- Reduce more potential overallocs in member `_current_metrics` by switching `std::vector<std::vector<metric_function>>` to `std::vector<chunked_fifo<metric_function>>`
- Reduce more potential overallocs in `values_copy::values` by changing `metrics_values` to a `std::deque` 

Changes from force-push `be869b7`:
- Commit https://github.com/scylladb/seastar/commit/3e9321cb46d30adf0d12ff24e401d718e619de0b added a default ctor to `chunked_fifo::basic_iterator`, so this force-push dropped the commit in this patch that added the same default ctor.

Changes from force-push `8a06f57`:
- Changed `chunked_fifo` to a `std::deque`

### CODE REVIEW

**Comment by amnonh** [comment]
The change looks safe LGTM

**Review by nyh — APPROVED** [review]
Looks good to me, although I would have preferred it to be split to two separate patches (one improving chunked_fifo, the other the metrics).

File: include/seastar/core/chunked_fifo.hh
Comment (tchaikov) [reply include/seastar/core/chunked_fifo.hh:149]: the reason we need to have an uninitialized iterator is that we need to create an empty `context` at https://github.com/scylladb/seastar/blob/e908cfce2d4488976f3b91906954f435aaa99238/src/core/scollectd.cc#L418 , and the `context` has an `seastar::metrics::impl::metric_metadata_fifo::iterator` after this change. without a default ctor, `make_lw_shared<context>()` won't compile.

Comment (tchaikov) [reply include/seastar/core/chunked_fifo.hh:149]: i concur with @nyh that it'd be better to explain why we need a default ctor for the `basic_iterator`. as it might be otherwise not self-explained.

Comment (nyh) [reply include/seastar/core/chunked_fifo.hh:149]: Yes, I guessed that some other code must have used an uninitialized iterator and initialize it later, and since this works with std::vector (and I didn't check, probably also with std::dequeue) then it should also work with chunked_fifo, so it's a good change.

Comment (NyaliaLui) [reply include/seastar/core/chunked_fifo.hh:149]: Yes, I added the default ctor to the iterator because of compilation issues with `make_lw_shared<context>()`. I will split up the commits and explain that further in the commit message.

**Comment by xemul** [comment]
Isn't the same jumbo-vector problem stands for `_current_metrics` and, consequentially, for inhabitants of `styruct values_copy`? @NyaliaLui

**Comment by NyaliaLui** [comment]
> Isn't the same jumbo-vector problem stands for `_current_metrics` and, consequentially, for inhabitants of `styruct values_copy`? @NyaliaLui

Apologies for late reply, I see how `_current_metrics` could overalloc but I don't see how `values_copy` are at risk. Could you explain a bit more?

Redpanda ran into this overalloc only with `metric_metadata_vector` so I changed it to `metric_metadata_fifo`. @xemul

File: include/seastar/core/metrics_api.hh
Comment (avikivity) [inline include/seastar/core/metrics_api.hh:327]: Isn't std::deque more suitable here?

**Comment by NyaliaLui** [comment]
> > Isn't the same jumbo-vector problem stands for `_current_metrics` and, consequentially, for inhabitants of `styruct values_copy`? @NyaliaLui
> 
> Apologies for late reply, I see how `_current_metrics` could overalloc but I don't see how `values_copy` are at risk. Could you explain a bit more?
> 
> Redpanda ran into this overalloc only with `metric_metadata_vector` so I changed it to `metric_metadata_fifo`. @xemul

I see `values_copy::values` overallocing since it gets the same size as `_current_metrics` in `src/core/metrics.cc`:
```cpp
foreign_ptr<values_reference> get_values() {
    shared_ptr<values_copy> res_ref = ::seastar::make_shared<values_copy>();
    auto& res = *(res_ref.get());
    auto& mv = res.values;
    res.metadata = get_local_impl()->metadata();
    auto & functions = get_local_impl()->functions();
    mv.reserve(functions.size()); // Potentially here
    for (auto&& i : functions) {
        value_vector values;
        values.reserve(i.size());
        for (auto&& v : i) {
            values.emplace_back(v());
        }
        mv.emplace_back(std::move(values));
    }
    return res_ref;
}
```

File: include/seastar/core/metrics_api.hh
Comment (NyaliaLui) [reply include/seastar/core/metrics_api.hh:327]: I argue for `chunked_fifo` because `std::deque` still does some contiguous allocations since it uses chunks of arrays. The problem that we were trying to address here was an overallocation issue on contiguous allocations.

May I ask, what is your concern? @avikivity

Comment (NyaliaLui) [reply include/seastar/core/metrics_api.hh:327]: > I argue for `chunked_fifo` because `std::deque` still does some contiguous allocations since it uses chunks of arrays. The problem that we were trying to address here was an overallocation issue on contiguous allocations.
> 
> May I ask, what is your concern? @avikivity

After posting I realized that `std::deque` is a good substitute for `values_copy::values` because it needs random access at locations [scollectd.cc:429](https://github.com/scylladb/seastar/blob/28daa8f50ccd5737b1779a2656d4330473d53d59/src/core/scollectd.cc#L429), [scollectd.cc:450](https://github.com/scylladb/seastar/blob/28daa8f50ccd5737b1779a2656d4330473d53d59/src/core/scollectd.cc#L450), [scollectd.cc:471](https://github.com/scylladb/seastar/blob/28daa8f50ccd5737b1779a2656d4330473d53d59/src/core/scollectd.cc#L471) and `chunked_fifo` does not support random access.

Comment (avikivity) [reply include/seastar/core/metrics_api.hh:328]: Mostly that std::queue is a standard type and more similar to std::vector.

Comment (NyaliaLui) [reply include/seastar/core/metrics_api.hh:328]: > Mostly that std::queue is a standard type and more similar to std::vector.

Thank you for the response. I am a little confused though, are there more items to discuss? cc: @avikivity

Comment (NyaliaLui) [reply include/seastar/core/metrics_api.hh:328]: > Mostly that std::queue is a standard type and more similar to std::vector.

These types are similar and they do have a design/interface that is unique compared to `chunked_fifo`. However, the variables `_current_metrics` and `metric_metadata_vector` do not seem to rely on those differences. A `chucked_fifo` also decreases the potential of memory fragmentation. For these reasons, I believe a `chunked_fifo` is OK.

I also believe a `std::deque` is OK for those places that could use improved memory management but still need the interfaces from `std`,  (e.g., `values_copy::values`).

**Comment by NyaliaLui** [comment]
@nyh @tchaikov I resolved conflicts and this patch is ready for re-review again.

**Comment by NyaliaLui** [comment]
@tchaikov @avikivity @amnonh 
Wondering if any other changes are needed here?

Also, @nyh it looks like github glitched and did not dismiss your review from July 16th

**Review by nyh — APPROVED** [review]
The patch looks ok to me, but I think the commit messages can be improved. In particular, both commit messages talk about "over allocations", but it isn't clear to me what this means. std::vector doesn't allocate **more** than seastar::chunked_fifo. I guess by "over allocation" you meant "oversized allocation", i.e., a single contiguous allocation that is very big, and Seastar warns about because in highly-fragmented memory it may become impossible to do this allocation and it will crash the application.

In general, it is always good to start by opening a github issue describing a **problem** (e.g., did you see an "oversized allocation" warning? did you run out of memory? In what case - did you add a lot of metrics? Or what?), and then your commit message can say it solves that issue (give a short summary and then say "Fixes #nnn").




@amnonh please review too, you wrote most of the code that this patch modifies.

**Review by tchaikov — APPROVED** [review]
lgtm

**Comment by NyaliaLui** [comment]
@nyh What you say makes sense, I created https://github.com/scylladb/seastar/issues/1844, linked this PR to that issue, and amended the PR description. It seems that changing the commit message is a non-blocker, so I'll leave that alone for now.

**Comment by tchaikov** [comment]
@scylladb/seastar-maint hello maintainers, could you help merge this change?

File: include/seastar/core/metrics_api.hh
Comment (avikivity) [reply include/seastar/core/metrics_api.hh:328]: I think deque is more suitable, chunked_fifo is a specialized type while deque is general-purpose.

Comment (NyaliaLui) [reply include/seastar/core/metrics_api.hh:328]: > I think deque is more suitable, chunked_fifo is a specialized type while deque is general-purpose.

@avikivity I am a little confused, are you requesting changes here?

Comment (avikivity) [reply include/seastar/core/metrics_api.hh:328]: Yes, please switch to deque.

Comment (NyaliaLui) [reply include/seastar/core/metrics_api.hh]: Made the change.

**Comment by NyaliaLui** [comment]
Pinging in case this got lost somewhere. CC @avikivity @tchaikov @xemul 

All requested changes are submitted.

**Pull request merged**
